### PR TITLE
 Conflict Confirmation When AutoSaving

### PR DESCRIPTION
**Pull request by DevIos01**
This PR adds support to #1247 after the game mechanic and testing are done (logs will be removed after approval)

Logic:
image

Screenshot:
image

### CODE REVIEW

File: src/components/big-interactive-pages/editor.tsx [hunk 1]
```diff
@@ -145,6 +163,38 @@ export default function Editor({ persistenceState, cookies }: EditorProps) {
...
+			}
+		};
+		channel.postMessage({ type: 'SESSION_UPDATE', sessionId });
+		return () => channel.close();
+	}, [sessionId]);
+
+	useEffect(() => {
+
+		const handleUnload = () => {
+			localStorage.removeItem('lastSavedSessionId');
```
Comment (grymmy) [inline src/components/big-interactive-pages/editor.tsx:183]: why is it important to delete this local session id on unload?  why is this necessary at all?

**Review by grymmy — CHANGES_REQUESTED** [review]

File: src/components/big-interactive-pages/editor.tsx [hunk 2]
```diff
@@ -145,6 +163,38 @@ export default function Editor({ persistenceState, cookies }: EditorProps) {
 	const screenContainer = useRef<HTMLDivElement>(null);
 	const screenControls = useRef<HTMLDivElement>(null);
 
+	const [sessionId] = useState(nanoid());
+
+	useEffect(() => {
```
Comment (grymmy) [inline src/components/big-interactive-pages/editor.tsx:168]: It is true that all this does detect new sessions?

File: src/components/big-interactive-pages/editor.tsx [hunk 3]
```diff
@@ -66,9 +68,23 @@ let lastSavePromise = Promise.resolve();
 let saveQueueSize = 0;
 export const saveGame = debounce(
 	800,
-	(persistenceState: Signal<PersistenceState>, code: string) => {
+	(persistenceState: Signal<PersistenceState>, code: string, sessionId: string) => {
 		const doSave = async () => {
 			const attemptSaveGame = async () => {
+				const game = (persistenceState.value.kind === 'PERSISTED' && persistenceState.value.game !== 'LOADING') ? persistenceState.value.game : null;
+				if (!game) return false;
+				const lastSavedSessionInfo = localStorage.getItem('lastSavedSessionId');
```
Comment (grymmy) [inline src/components/big-interactive-pages/editor.tsx:76]: Please modify all usages of the 'lastSavedSessionId' string literal to a constant

File: src/components/big-interactive-pages/editor.tsx (hunk 1 above)
Comment (DevIos01) [reply src/components/big-interactive-pages/editor.tsx:183]: The **lastSavedSessionId** is removed on page unload to ensure that each session starts fresh, preventing the website from mistakenly displaying an overwrite warning when users return to the editor.

Comment (DevIos01) [reply src/components/big-interactive-pages/editor.tsx:183]: it's replicate how SessionStorage works, it destory on session / tab end, problem with SessionStorage on our case is that it doesn't communicate between tabs, so we need used LocalStorage with this manual remove on unload.

File: src/components/big-interactive-pages/editor.tsx [hunk 4]
```diff
@@ -66,9 +68,23 @@ let lastSavePromise = Promise.resolve();
...
+				const lastSavedData = lastSavedSessionInfo ? JSON.parse(lastSavedSessionInfo) : null;
+				const lastSavedSessionId = lastSavedData ? lastSavedData.sessionId : null;
+				const lastSavedGameId = lastSavedData ? lastSavedData.gameId : null;
+				
+				console.log('Attempting to save. Current session:', sessionId, 'Last saved session:', lastSavedSessionId);
+				
+				if (lastSavedGameId === game.id && lastSavedSessionId !== sessionId) {
+					showSaveConflictModal.value = true;
+					continueSaving.value = false;
+					return false;
```
Comment (grymmy) [inline src/components/big-interactive-pages/editor.tsx:86]: TBH, I think a better, high-level pattern here would be the following:

```
if(shouldShowConflict(cxt)) {
  showConflict(cxt);
} else {
  if (!save(cxt)) {
    // show error
  }
}
```

Instead, what we're working with here is:
```
if (!save(cxt)) {
  // Figure out why saving failed, it left us a breadcrumb multivariable trail of clues
  if (cxt.continueSaving) { // logic } else if (cxt.showSaveConflictModal) { //logic }....
```

File: src/components/big-interactive-pages/editor.tsx (hunk 2 above)
Comment (DevIos01) [reply src/components/big-interactive-pages/editor.tsx:168]: yes currently all it does is communicate with session on each tabs aka broadcasting the message to all tabs. (mostly used for testing purposes looking at it now)

Comment (grymmy) [reply src/components/big-interactive-pages/editor.tsx:168]: can this be scoped out and separated to make the logic of how the editor is init'd more concise?

Comment (grymmy) [reply src/components/big-interactive-pages/editor.tsx:168]: i.e. useEffect(logNewSessions)

File: src/components/big-interactive-pages/editor.tsx [hunk 5]
```diff
@@ -147,6 +174,38 @@ export default function Editor({ persistenceState, cookies }: EditorProps) {
 	const screenContainer = useRef<HTMLDivElement>(null);
 	const screenControls = useRef<HTMLDivElement>(null);
 
+	const [sessionId] = useState(nanoid());
+
+	useEffect(() => {
+		console.log('New session started with ID:', sessionId);
```
Comment (JosiasAurel) [inline src/components/big-interactive-pages/editor.tsx:180]: let's remove the console logs

**Review by JosiasAurel — APPROVED** [review]
